### PR TITLE
enable Context Parallel

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #592

**Summary**
This PR adds DTensor Context Parallel (https://github.com/pytorch/pytorch/pull/131351) option to torchtitan. 

**Instructions**
Find the `context_parallel_degree` option in the `experimental` section in `.toml` file, assign the desired Context Parallel degree (i.e. the # of GPUs along the Context Parallel dimension). 

**Benchmark Settings**
The benchmark uses `train_configs/llama3_8b.toml` with extra changes as below:

- norm_type = "fused_rmsnorm"
- mode = 'full' (using full activation checkpoint allows us to find the longest possible input sequence length on a device)

**Benchmark Results**
1. *max seq_len scalability*
The first set of benchmark shows that with CP, adding more GPUs on the CP dimension allows longer input seq_len and this growth is linearly scalable. We use FSDP to shard the model and fix `data_parallel_shard_degree` to be 8, and we use 8, 16, 32, 64 H100 GPUs (i.e. CP degree=1,2,4,8) to measure the longest possible input seq_len w/o GPU OOM. We adjust the CP degree by modifying the `context_parallel_degree` in the `experimental` section in `.toml` file.
image

*As Context Parallel degrees goes up (1, 2, 4, 8), the max seq_len also increases (32k, 80k, 144k, 300k)*

image

*While increasing Context Parallel degree increases the max seq_len, it also decreases the WPS on each device. This is the price paid for longer input*

image

*However, increasing CP degree doesn't significantly affect MFU.*

2. *loss curve convergence*
The second set of benchmark proves that our implementation is correct by showing the loss curve converges just as FSDP + TP. This experiment uses 64 H100 GPUs and fix TP degree and local batch size to 8 (`tensor_parallel_degree=8`, `batch_size=8`). As we increase CP degree, we decrease DP degree to make sure that `CP * DP = 8`. Besides, we also make sure `seq_len` increases such that `seq_len = 8192 * CP`.

image

*The training loss curve matches among the 4 combination of parallelisms: DP=8; DP=4, CP=2; DP=2, CP=4; CP=8 (TP is fixed to 8).*

Screenshot 2024-10-07 at 12 42 37 PM

*The only observed difference is among the warm-up steps. This is due to our pick of the number of steps and picking an appropriate number of warm-up steps should smooth out this.*

3. *max seq_len vs. local WPS on a fixed number of GPUs*
The third set of benchmark gives users a guide on what CP degree to choose and what WPS to expect on a specific set of devices (e.g. number of devices, device types, etc). Here we use H100 as an example but it is expected to observer similar trade-off curves on other hardware. 

image

*Increasing CP degree allows longer input sequence length at the cost of reducing WPS on each device. The exception is pure CP -- the seq_len increase is very limited compared to DP=2, CP=4 on 8 GPUs. This is due to our implementation, and we're working on enabling performant CP to support ultra-long input sequence.*

TODO: 
1. An option for CP tensor rotate implementation between all-to-all and all-gather will be added once https://github.com/pytorch/pytorch/pull/132820 is landed. 
2. Investigate the limitation in pure CP, to enable ultra-long input sequence.

### CODE REVIEW

File: torchtitan/models/llama/model.py [hunk 1]
```diff
@@ -415,8 +415,8 @@ def _precompute_freqs_cis(self) -> torch.Tensor:
         return precompute_freqs_cis(
             self.model_args.dim // self.model_args.n_heads,
             # Need to compute until at least the max token limit for generation
-            # (use 2x max sequence length to be safe)
-            self.model_args.max_seq_len * 2,
+            # Note: removed the 2x relaxing in CP enablement
+            self.model_args.max_seq_len,
```
Comment (fegin) [inline torchtitan/models/llama/model.py:420]: cc., @tianyu-l Want to understand is this okay?

For a general use case, we can also expand the CP to support `stride`-like feature.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 2]
```diff
@@ -72,13 +74,18 @@ def parallelize_llama(
...
         if parallel_dims.dp_shard_enabled:
             if parallel_dims.dp_replicate_enabled:
                 dp_mesh = world_mesh["dp_replicate", "dp_shard"]
             else:
                 dp_mesh = world_mesh["dp"]
 
+            if parallel_dims.cp_enabled:
+                dp_dim_names = dp_mesh.mesh_dim_names
+                assert isinstance(dp_dim_names, Tuple)
+                dp_mesh = world_mesh[(*dp_dim_names, "cp")]._flatten()
```
Comment (fegin) [inline torchtitan/parallelisms/parallelize_llama.py]: I remember we want to initialize all the PG in the very beginning. Can we move this to `parallel_dims.py` and use `mesh_dim_name` to rename it to `dp_shard_cp`?

File: torchtitan/models/llama/model.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/llama/model.py:420]: Could you please elaborate a bit on why this change was needed by CP?

Comment (fegin) [reply torchtitan/models/llama/model.py:420]: @tianyu-l CP parallelize on the sequence dimension, anything related to the sequence dimension needs to be shard. So freqs_cis is the positional embedding and is required to be sharded according to the sequence length. So it is easier to support CP if everything has the same sequence length.

File: test_runner.py [hunk 3]
```diff
@@ -306,6 +306,29 @@ def build_test_list():
...
+                [
+                    "--training.data_parallel_shard_degree=2",
+                    "--training.data_parallel_replicate_degree=2",
+                    "--experimental.context_parallel_degree=2",
+                ]
+            ],
+            "HSDP+CP",
+            "hsdp+cp",
+            ngpu=8,
+        ),
```
Comment (tianyu-l) [inline test_runner.py:331]: If FSDP + TP + CP works, we should create one more test.

File: torchtitan/models/llama/model.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/llama/model.py:420]: Sounds reasonable to me. @awgu to confirm this is OK.

Also we need to add a note in docs/composability.md to clarify why this (model change) is needed. It can be addressed in a separate PR; in that case please create issue / leave TODO.

File: train.py [hunk 4]
```diff
@@ -268,18 +313,24 @@ def loss_fn(pred, labels):
             data_load_start = time.perf_counter()
             batch = next(data_iterator)
             input_ids, labels = batch
-            ntokens_since_last_log += labels.numel()
+            ntokens_since_last_log += labels.numel() // parallel_dims.cp
```
Comment (tianyu-l) [inline train.py]: instead of accounting it here, we should account it at https://github.com/pytorch/torchtitan/blob/main/train.py#L351
and modify `model_parallel_size` to `non_data_parallel_size`

File: train.py [hunk 5]
```diff
@@ -28,17 +34,54 @@
 )
 from torchtitan.profiling import maybe_enable_memory_snapshot, maybe_enable_profiling
 
+try:
+    from torch.distributed.tensor.experimental import context_parallel
+except ImportError:
+    print(
+        f"PyTorch version {torch.__version__} does not include the experimental "
+        "Context Parallel API. Please update to a newer version."
+    )
```
Comment (tianyu-l) [inline train.py]: let's move this import to local, rather than global

File: train.py [hunk 6]
```diff
@@ -8,9 +8,15 @@
...
 from datetime import timedelta
+from functools import partial
+
+from typing import List, Optional, Set
 
 import torch
+
+from torch.distributed.device_mesh import DeviceMesh
 from torch.distributed.elastic.multiprocessing.errors import record
+from torch.nn.attention import sdpa_kernel, SDPBackend
```
Comment (tianyu-l) [inline train.py]: ditto: please make this import local, or e.g. just use `torch.nn.attention.sdpa_kernel` if it appears only once

File: train.py [hunk 7]
```diff
@@ -28,17 +34,54 @@
...
+    if cp_mesh is not None:
+        context_parallel_ctx = partial(context_parallel, mesh=cp_mesh)
 
-def get_train_context(enable_loss_parallel: bool, enable_compiled_autograd: bool):
     @contextlib.contextmanager
-    def context():
+    def context(
+        cp_buffers: List[torch.Tensor],
+        cp_seq_dims: List[int],
+        cp_no_restore_buffers: Set[torch.Tensor],
```
Comment (tianyu-l) [inline train.py]: Instead of passing CP related args to the general `context()`, we should create a "context parallel" context as the arg. And later on call `stack.enter_context()` to enter this context. We can create a separate function to generate this context given the CP related args. This function should be put into another file, e.g.`torchtitan/parallelisms/utils.py`.

File: train.py [hunk 8]
```diff
@@ -8,9 +8,15 @@
 import os
 import time
 from datetime import timedelta
+from functools import partial
+
+from typing import List, Optional, Set
```
Comment (tianyu-l) [inline train.py]: Let's put these extra imports, including `partial`, `DeviceMesh` to the separate util file, as commented below. This can help `train.py` stay as clean as possible.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 9]
```diff
@@ -72,36 +73,44 @@ def parallelize_llama(
...
+    if (
+        parallel_dims.dp_shard_enabled
+    ):  # apply FSDP or HSDP, potentially with Context Parallel
+        dp_mesh_dim_names = (
+            ("dp_replicate", "dp_shard")
+            if parallel_dims.dp_replicate_enabled
+            else ("dp",)
+        )
+        dp_mesh = (
+            world_mesh[(*dp_mesh_dim_names, "cp")]._flatten("dp_cp")
```
Comment (tianyu-l) [inline torchtitan/parallelisms/parallelize_llama.py:91]: We need to document this design in torchtitan, e.g. in the file `docs/context_parallel.md`. This can be addressed in another PR.

**Review by tianyu-l — COMMENTED** [review]
Thanks for enabling this. Had some suggestions, including how to keep `train.py` clean.

Also if possible, let's add the perf / loss validation progress in the PR summary.

File: train.py (hunk 8 above)
Comment (XilunWu) [reply train.py]: I'm not sure if this helps keep the file clean. Suppose we put `List` in `torchtitan/utils.py`, then we need to use `List`, we instead use `utils.List` which does not look straightforward to users (this may look like a custom user-defined type).

Let me think of a better way if there's one.

File: train.py (hunk 7 above)
Comment (XilunWu) [reply train.py:58]: I'm not sure if creating CP context on demand is a good idea because of the overhead. Besides, the computation in ring attention is very regular and repeated so re-using the context looks reasonable to me. cc @fegin

File: train.py (hunk 8 above)
Comment (fegin) [reply train.py]: I think what @tianyu-l  means is that if we move the new logic of `get_train_context` to `parallelisms/utils.py` file, these new import will also be removed from `train.py`. Otherwise, indirect importing is not a good idea.

Comment (XilunWu) [reply train.py]: I moved `get_train_context` to `utils.py`. Let me know if this is not okay. cc @fegin @tianyu-l

**Comment by awgu** [comment]
The loss curves look a bit strange to me.

File: torchtitan/utils.py [hunk 10]
```diff
@@ -101,6 +102,55 @@ def run(self, step_count):
...
+            "Context Parallel API. Please update to a newer version."
+        )
+
+    return context_parallel(
+        cp_mesh,
+        buffers=cp_buffers,
+        buffer_seq_dims=cp_seq_dims,
+        no_restore_buffers=cp_no_restore_buffers,
+    )
+
```
Comment (tianyu-l) [inline torchtitan/utils.py:125]: nit: two empty lines in between, not sure if it's covered by `pre-commit`.
btw it seems linting test failed.

File: test_runner.py [hunk 11]
```diff
@@ -306,6 +306,41 @@ def build_test_list():
...
+                [
+                    "--training.data_parallel_shard_degree=2",
+                    "--training.tensor_parallel_degree=2",
+                    "--experimental.context_parallel_degree=2",
+                ]
+            ],
+            "FSDP+TP+CP",
+            "fsdp+tp+cp",
+            ngpu=8,
+        ),
```
Comment (tianyu-l) [inline test_runner.py:343]: Question: Looks like FSDP/HSDP + TP + CP is working. How about PP? We can also mention progress in the .md doc later.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 12]
```diff
@@ -72,36 +73,44 @@ def parallelize_llama(
...
-                tp_enabled=parallel_dims.tp_enabled,
-                pp_enabled=parallel_dims.pp_enabled,
-            )
-            if parallel_dims.dp_replicate_enabled:
-                logger.info("Applied HSDP to the model")
-            else:
-                logger.info("Applied FSDP to the model")
+    if (
+        parallel_dims.dp_shard_enabled
+    ):  # apply FSDP or HSDP, potentially with Context Parallel
```
Comment (tianyu-l) [inline torchtitan/parallelisms/parallelize_llama.py:78]: nit: can we remove the parenthesis?

File: torchtitan/parallelisms/parallelize_llama.py (hunk 9 above)
Comment (tianyu-l) [reply torchtitan/parallelisms/parallelize_llama.py:91]: `_flatten` seems called twice, another one in `parallel_dims.py`. I wonder how this API works?

Comment (tianyu-l) [reply torchtitan/parallelisms/parallelize_llama.py:91]: another question is: given the current implementation, does it mean DP and CP have to be adjacent to each other in the device mesh? E.g. it seems we can't do [TP, CP, PP, DP] (from inner to outer) as in Llama 3.1 paper. Is that correct?

**Review by tianyu-l — COMMENTED** [review]
Looks awesome! Still have a few questions, but I think it's mostly ready.

**Comment by XilunWu** [comment]
@awgu thanks for catching it! I forgot to paste the 5000-step loss curve screenshot but a zoom-in one for warm-up steps. Updated.

File: test_runner.py (hunk 11 above)
Comment (XilunWu) [reply test_runner.py:343]: Right, the next step is to test 4D/5D (w/ PP and HSDP)

File: torchtitan/parallelisms/parallelize_llama.py (hunk 12 above)
Comment (XilunWu) [reply torchtitan/parallelisms/parallelize_llama.py:78]: unfortunately no, this is due to the comment following. putting them in one line violates the code length requirement.

**Comment by tianyu-l** [comment]
> While increasing Context Parallel degree increases the max seq_len, it also decreases the WPS on each device. This is the price paid for longer input. However, increasing CP degree doesn't significantly affect MFU.

By looking into how token/sec and MFU change when increasing `seq_len`, here's my explanation to this phenomenon.

There are two parts of flops computation
https://github.com/pytorch/torchtitan/blob/main/torchtitan/utils.py#L150
- (I) matmul scales linearly with number of tokens
- (II) attention scales quadratically with number of tokens

When doubling seq_len together with CP degree (also doubling number of GPUs since other degrees are fixed)
- (I) matmul computation doubles; (II) attention computation quadruples.
- Even CP doesn’t bring any overhead (e.g. extra comm exposure), we shoudn’t expect the same WPS per GPU because computation resource is doubled, but computation load is between 2x – 4x. The throughput per GPU should be between 0.5x and 1x, assuming perfect scaling.
- The MFU change depends on the actual config, more specifically between the ratio of (I) and (II). If (I) dominates, the MFU/WPS ratio wouldn’t change as we increase seq_len; if (II) dominates, the MFU/WPS ratio would increase linearly with seq_len. For 8B and seq_len around 32k/64k, (II) is larger than (I) so that’s why we are seeing WPS drop, but MFU stays (because the ratio increases).

My conclusion is that the perf scaling we observed makes sense and look good to me.

File: torchtitan/parallelisms/parallelize_llama.py (hunk 12 above)
Comment (tianyu-l) [reply torchtitan/parallelisms/parallelize_llama.py:78]: can we put the comment one above the if-statement?

File: torchtitan/parallelisms/parallelize_llama.py (hunk 9 above)
Comment (XilunWu) [reply torchtitan/parallelisms/parallelize_llama.py:91]: Regarding the mesh order, the current implementation of mesh flattening requires the flattened dimensions are contiguous which means `[TP, CP, PP, DP]` won't work in torchtitan.

Comment (XilunWu) [reply torchtitan/parallelisms/parallelize_llama.py:91]: For the first question on `_flatten`, it should actually be only called once. I have updated the call in `parallelize_llama.py` to a simple mesh access.

**Review by tianyu-l — APPROVED** [review]
LGTM! Thanks!

**Comment by XilunWu** [comment]
@tianyu-l actually, the 8 GPU test has a failure. This is the HSDP + CP case. I think the flatten logic has some bug when handling `mesh["dp_cp"]`. 

Flattening `mesh["dp_replicate", "dp_shard", "cp"]` into `mesh["dp_cp"]` is a workaround w/o actual cost. I think we can land this first with a TODO and put the `__get_item__` call back once it's fixed in DeviceMesh.

File: torchtitan/utils.py [hunk 13]
```diff
@@ -101,6 +102,57 @@ def run(self, step_count):
...
+            if enable_compiled_autograd:
+                stack.enter_context(
+                    torch._dynamo.utils.maybe_enable_compiled_autograd(True)
+                )
+
+            if cp_context is not None:
+                from torch.nn.attention import sdpa_kernel, SDPBackend
+
+                # currently we only support these two SDP backends.
+                # TODO (xilunwu): support cuDNN backend
```
Comment (tmm1) [inline torchtitan/utils.py:143]: Just curious if you recall what the blocker for `CUDNN_ATTENTION` is

Comment (tianyu-l) [reply torchtitan/utils.py:143]: @XilunWu

Comment (XilunWu) [reply torchtitan/utils.py:143]: Hi @tmm1 it's simply cudnn attention has a different op signature. I'm adding support now and should be able to have the PR draft out by next week.

Comment (xingchensong) [reply torchtitan/utils.py:143]: Hi @XilunWu, just curious. Is cuDNN backend supported now?

Comment (XilunWu) [reply torchtitan/utils.py:143]: @xingchensong Yes!! It's merged in https://github.com/pytorch/pytorch/pull/148537

cc @tmm1
